README.md
# 2PLSF - Two-Phase Locking with Starvation-Freedom

This repository contains the source code and benchmarks of the paper:

Pedro Ramalhete, Andreia Correia, Pascal Felber, [2PLSF - Two-Phase Locking with Starvation-Freedom](draft not yet available), PPoPP 2023

The source code the 2PLSF is located in stms/2PLSF.hpp and no other file is required to use this concurrency control in your work.

## STMs
This repository contains the following STM implementations in the stms/ folder:

- TinySTM.hpp and tinystm/			&nbsp;&nbsp; Wrapper of TinySTM with the LSA concurrency control
- TL2ORig.hpp and tl2-x86/			&nbsp;&nbsp; Wrapper of the original TL2 concurrency control implementation
- zardoshti/orec_eager_wrap.hpp		&nbsp;&nbsp; Wrapper of the OREC eager STM
- zardoshti/orec_lazy_wrap.hpp		&nbsp;&nbsp; Wrapper of the OREC lazy STM
- zardoshti/tl2_wrap.hpp			&nbsp;&nbsp; Wrapper of the TL2 implementation by Zardoshti et al
- zardoshti/tlrw_eager_wrap.hpp		&nbsp;&nbsp; Wrapper of the TLRW implementation by Zardoshti et al
- OneFileWF.hpp						&nbsp;&nbsp; OneFile Wait-Free STM
- 2PLSF.hpp							&nbsp;&nbsp; 2PLSF Software Transactional Memory
- 2PLUndo.hpp						&nbsp;&nbsp; 2PL with Undo and non-scalable reader-writer lock
- 2PLUndoDist.hpp					&nbsp;&nbsp; 2PL with Undo and scalable reader-writer lock

We use wrapper classes to integrate some of the existing STMs with our benchmarks.



## Data structure benchmarks
The folder graphs/ contains the benchmarks for data structures.

### Build data structure benchmarks
To build you must first build the TinySTM:

	cd stms/tinystm
	make

then build TL2

	cd ../tl2-x86
	make
	
then build the data structures benchmarks:

	cd ../../graphs
	make

If you don't have g++-10 then edit the Makefile in the folder above and replace CXX with the compiler you wish to use.

### Run data structure benchmarks
To run the benchmarks you can go into the graphs/ folder and run the python script run-all.py

	python2 run-all.py


## DBx1000 benchmarks
The source code for the DBx1000 benchmark was taken from https://github.com/yxymit/DBx1000
We fixed a couple of issues but the inner workings stayed the same. We added the 2PLSF concurrency control, with the file stms/2PLSF.hpp being needed in addition to the files added in DBx1000/

We have also disabled the abort buffer as it will not work properly with 2PLSF because 2PLSF waits for a specific thread/transaction to complete.
You can check which settings to use by looking at run-dbx1000.py or executing this script to generate YCSB results for the currently selected Concurrency Control (CC).
To chose a different CC, open DBx1000/config.h and replace CC_ALG with one of the other names: WAIT_DIE, NO_WAIT, DL_DETECT, SILO, HEKATON, TICTOC, TWO_PL_SF

2PLSF will disregard the ABORT_PENALTY because it uses its own restarting scheme.

We also added a "continue" in the main loop which causes NO_WAIT to get stuck due to permanent conflicts aborting themselves. This happens because NO_WAIT is prone to live-lock, it's just that the DBx1000 benchmarks were originally made so that this would not be seen.

Keep in mind that this particular implementation of DBx1000 does no exercise the indexing data structure, as the index is a sequential hashmap. In other words, the YCSB of this benchmark is *not* the true YCSB because there are no record insertions nor deletions, only record updates of pre-existing records.
For slightly more realistic YCSB benchmarks, take a look at setbench: https://gitlab.com/trbot86/setbench/-/wikis/home

After editing the config.h, build with

	make -j
and run with
	
	./rundb
or

	./run-dbx1000.py



## Figures
After running the benchmarks, it's possible to obtain the figures from the paper by going in the graphs/plots folder and running ./plot-all.sh
However, the DBx1000 results need to be copied one by one to the appropriate files in the graphs/data/cervino/ folder, as this benchmark is not easily automated.
The plot-all.sh script requires a recent gnuplot and a few other tools to run correctly, like ps2pdf.


## How to use 2PLSF in a different benchmark
To use 2PLSF in your own application or benchmark, include the 2PLSF.hpp header in one of your .cpp files. You will not need any other file.
If for some reason you need to include it from multiple .cpp files, then #define INCLUDED_FROM_MULTIPLE_CPP before the #include<2PLSF.hpp>. Do this from all the .cpp files except one, otherwise you will see multiple declarations of the globals.
Alternatively, you can move the last block of code in 2PLSF.hpp to one of your .cpp files, and then include from all the .cpp files without the extra #define.

You can start transactions by passing a lambda to twoplsf::transaction() with txType=TX_IS_UPDATE for write-transactions and txType=TX_IS_READ for read-transactions. Read accesses should be done through tmtype::pload() and write access through tmtype::pstore().


For further questions go to 
http://www.github.com/pramalhe/2PLSF.git

stms/ChangeLog
[unreleased]

2014-12-01  Patrick Marlier <patrick.marlier@unine.ch>
	* Remove hybrid implementation (ASF)

[1.0.5 release]

2013-01-14  Pascal Felber  <pascal.felber@unine.ch>
	* Fixed bugs in garbage collection.
	* Added module that for to commit in order (mod_order)
	* Added stm_get_specific_tx and stm_set_specific_tx for mod_order.

[1.0.4 release]

2012-10-23  Pascal Felber  <pascal.felber@unine.ch>
	* Added MIT license (dual license).

2012-03-27  Patrick Marlier <patrick.marlier@unine.ch>
	* Split stm.c file for the different strategies (wb-etl/wb-ctl/wt).

[1.0.3 release]

2011-11-17  Patrick Marlier <patrick.marlier@unine.ch>
	* Fixed stack saving and minor changes for DTMC (tanger).

2011-10-20  Patrick Marlier <patrick.marlier@unine.ch>
	* Fixed a bug in tanger_stm_realloc (reported by Luke Dalessandro).

2011-02-15  Patrick Marlier <patrick.marlier@unine.ch>
	* Fixed a bug in stm_rollback (reported by Rahul Gayatri).

[1.0.2 release]

2011-02-07  Patrick Marlier <patrick.marlier@unine.ch>
	* Minor fixes.
	* Added stm_get_attribute_tx() for Event Processing.
	* Added transactional_safe new/delete operators for GCC-TM.

2010-08-10  Patrick Marlier <patrick.marlier@unine.ch>
	* ABI uses WRITE_THROUGH by default.
	* Added irrevocability management for hybrid ASF mode.
	* Disable inlining for tanger load/store due to a LLVM bug.
	* Fixed a bug in gc.c (thanks to Walther Maldonado).
	* Fixed a bug in mod_log.c (thanks to Adam Cozzette).

2010-07-08  Patrick Marlier <patrick.marlier@unine.ch>
	* Fixed a bug in stm_commit with CTL (thanks to Srđan Stipić).

2010-07-07  Pascal Felber  <pascal.felber@unine.ch>
	* Fixed a bug in stm_set_irrevoable when using abort handler.

[1.0.1 release]

2010-06-02  Patrick Marlier <patrick.marlier@unine.ch>
	* Added Hybrid TM using AMD ASF.
	* Added ABI compatability with Intel STM C Compiler, GCC with TM
	support and DTMC.
	* Removed outdated Tanger support (now DTMC in ABI).

2010-04-29  Pascal Felber  <pascal.felber@unine.ch>
	* Added mod_ab module (atomic block statistics).

[1.0.0 release]

2010-02-22  Pascal Felber  <pascal.felber@unine.ch>
	* Various fixes and cosmetic changes.

2010-01-15  Pascal Felber  <pascal.felber@unine.ch>
	* Changed rollover code.
	* Added quiescence and serial-irrevocable mode.

2009-12-18  Pascal Felber  <pascal.felber@unine.ch>
	* Added commit/abort callbacks.
	* Various fixes.

2009-12-15  Pascal Felber  <pascal.felber@unine.ch>
	* Updated irrevocability to work with modular contention manager.

2009-11-29  Pascal Felber  <pascal.felber@unine.ch>
	* Updated modules to avoid name conflicts.

2009-11-20  Pascal Felber  <pascal.felber@unine.ch>
	* Changed signature of stm_start().
	* Added irrevocability.
	* Added ABI for intel compiler and tm-gcc.
	* Fixed a bug in mod_log.

2009-09-25  Pascal Felber  <pascal.felber@unine.ch>
	* Can disable alternation of insertion/removal in intset.

2009-06-10  Pascal Felber  <pascal.felber@unine.ch>
	* Added customizable (modular) contention manager.

2009-03-08  Pascal Felber  <pascal.felber@unine.ch>
	* Added some more statistics.
	* Added call to conflict callback upon validation.

2009-02-06  Pascal Felber  <pascal.felber@unine.ch>
	* Added version number.
	* Renamed mod_local module to mod_log and updated code.

2009-02-06  Pascal Felber  <pascal.felber@unine.ch>
	* Added abort reason as parameter of siglongjmp().

[0.9.9 release]

2009-01-29  Pascal Felber  <pascal.felber@unine.ch>
	* Updated TANGER wrappers.
	* Prepared new release.

2009-01-17  Pascal Felber  <pascal.felber@unine.ch>
	* Added content of README as doxygen documentation.
	* Added new version of TANGER wrappers (partial).
	* Added region load/store and support for misaligned accesses.
	* Bug fixes.

2009-01-04  Pascal Felber  <pascal.felber@unine.ch>
	* Documented APIs (doxygen comments in include files).
	* Cleaned up some APIs, changed parameters of stm_start().
	* CTL now spins while address is locked.

2008-12-06  Pascal Felber  <pascal.felber@unine.ch>
	* Cleaned up atomic operations and memory barriers.

2008-11-26  Pascal Felber  <pascal.felber@unine.ch>
	* Documented compile-time options in Makefile

2008-11-16  Pascal Felber  <pascal.felber@unine.ch>
	* Added support to read previous version of locked data.
	* Added support in Makefile for unifdef.

2008-07-22  Pascal Felber  <pascal.felber@unine.ch>
	* Added epoch-based memory allocator.
	* Added conflict tracking and callback.
	* Added skip list and refactored intset benchmark
	* Added mod_local module.
	* Added functions to control extensions and read the clock.
	* Added support for unit transactions in intset.
	* Many other changes and fixes.

[0.9.5 release]

2008-07-07  Pascal Felber  <pascal.felber@unine.ch>
	* Added support for the latest version of STAMP.
	* Removed TANGER wrappers (distributed separately).
	* Many modifications and fixes.

2008-05-26  Pascal Felber  <pascal.felber@unine.ch>
	* Added CTL.

2008-05-20  Pascal Felber  <pascal.felber@unine.ch>
	* Major API changes (no more transaction parameters, etc.).
	* Added unit transactions.
	* Many other changes and fixes.

[0.9.0 release]

2008-04-25  Pascal Felber  <pascal.felber@unine.ch>
	* Removed parameter to stm_new() => transaction cannot be on stack.

2008-03-27  Pascal Felber  <pascal.felber@unine.ch>
	* Added new module for coarse-grain statistics.
	* Added new internal statistics.
	* Added new support for reading configuration parameters.

2008-03-25  Pascal Felber  <pascal.felber@unine.ch>
	* Merged write-back and write-through in a single file.

2008-03-16  Pascal Felber  <pascal.felber@unine.ch>
	* Refactored dynamic memory management as a module.

2008-03-15  Pascal Felber  <pascal.felber@unine.ch>
	* Added support for modules.
	* Added sample mod_print module.


DBx1000/config.h
#ifndef _CONFIG_H_
#define _CONFIG_H_

/***********************************************/
// Simulation + Hardware
/***********************************************/
#define THREAD_CNT				    2
#define PART_CNT					1 
// each transaction only accesses 1 virtual partition. But the lock/ts manager and index are not aware of such partitioning. VIRTUAL_PART_CNT describes the request distribution and is only used to generate queries. For HSTORE, VIRTUAL_PART_CNT should be the same as PART_CNT.
#define VIRTUAL_PART_CNT			1
#define PAGE_SIZE					4096 
#define CL_SIZE						64
// CPU_FREQ is used to get accurate timing info 
#define CPU_FREQ 					2 	// in GHz/s

// # of transactions to run for warmup
#define WARMUP						0
// YCSB or TPCC
#define WORKLOAD YCSB
// print the transaction latency distribution
#define PRT_LAT_DISTR				false
#define STATS_ENABLE				true
#define TIME_ENABLE					true 

#define MEM_ALLIGN					8 

// [THREAD_ALLOC]
#define THREAD_ALLOC				false
#define THREAD_ARENA_SIZE			(1UL << 22) 
#define MEM_PAD 					true

// [PART_ALLOC] 
#define PART_ALLOC 					false
#define MEM_SIZE					(1UL << 30) 
#define NO_FREE						false

/***********************************************/
// Concurrency Control
/***********************************************/
// WAIT_DIE, NO_WAIT, DL_DETECT, TIMESTAMP, MVCC, HEKATON, HSTORE, OCC, VLL, TICTOC, SILO, TWO_PL_SF
// TODO TIMESTAMP does not work at this moment
#define CC_ALG                      TWO_PL_SF
#define ISOLATION_LEVEL 			SERIALIZABLE

// all transactions acquire tuples according to the primary key order.
#define KEY_ORDER					false
// transaction roll back changes after abort
#define ROLL_BACK					true
// per-row lock/ts management or central lock/ts management
#define CENTRAL_MAN					false
#define BUCKET_CNT					31
#define ABORT_PENALTY 				100000  // default is 100k
#define ABORT_BUFFER_SIZE			10
#define ABORT_BUFFER_ENABLE			false   // default is true
// [ INDEX ]
#define ENABLE_LATCH				false
#define CENTRAL_INDEX				false
#define CENTRAL_MANAGER 			false
#define INDEX_STRUCT				IDX_HASH
#define BTREE_ORDER 				16

// [DL_DETECT] 
#define DL_LOOP_DETECT				1000 	// 100 us
#define DL_LOOP_TRIAL				100	// 1 us
#define NO_DL						KEY_ORDER
#define TIMEOUT						1000000 // 1ms
// [TIMESTAMP]
#define TS_TWR						false
#define TS_ALLOC					TS_CAS
#define TS_BATCH_ALLOC				false
#define TS_BATCH_NUM				1
// [MVCC]
// when read/write history is longer than HIS_RECYCLE_LEN
// the history should be recycled.
//#define HIS_RECYCLE_LEN				10
//#define MAX_PRE_REQ					1024
//#define MAX_READ_REQ				1024
#define MIN_TS_INTVL				5000000 //5 ms. In nanoseconds
// [OCC]
#define MAX_WRITE_SET				10
#define PER_ROW_VALID				true
// [TICTOC]
#define WRITE_COPY_FORM				"data" // ptr or data
#define TICTOC_MV					false
#define WR_VALIDATION_SEPARATE		true
#define WRITE_PERMISSION_LOCK		false
#define ATOMIC_TIMESTAMP			"false"
// [TICTOC, SILO]
#define VALIDATION_LOCK				"no-wait" // no-wait or waiting
#define PRE_ABORT					"true"
#define ATOMIC_WORD					true
// [HSTORE]
// when set to true, hstore will not access the global timestamp.
// This is fine for single partition transactions. 
#define HSTORE_LOCAL_TS				false
// [VLL] 
#define TXN_QUEUE_SIZE_LIMIT		THREAD_CNT

/***********************************************/
// Logging
/***********************************************/
#define LOG_COMMAND					false
#define LOG_REDO					false
#define LOG_BATCH_TIME				10 // in ms

/***********************************************/
// Benchmark
/***********************************************/
// max number of rows touched per transaction
#define MAX_ROW_PER_TXN				64
#define QUERY_INTVL 				1UL
#define MAX_TXN_PER_PART 			100000 // default is 100k
#define FIRST_PART_LOCAL 			true
#define MAX_TUPLE_SIZE				1024 // in bytes
// ==== [YCSB] ====
#define INIT_PARALLELISM			1
#define SYNTH_TABLE_SIZE 			(1024 * 1024 * 10)  // default is 10M
#define ZIPF_THETA 					0.6
#define READ_PERC 					0.9
#define WRITE_PERC 					0.1
#define SCAN_PERC 					0
#define SCAN_LEN					20
#define PART_PER_TXN 				1
#define PERC_MULTI_PART				1
#define REQ_PER_QUERY				16
#define FIELD_PER_TUPLE				10
// ==== [TPCC] ====
// For large warehouse count, the tables do not fit in memory
// small tpcc schemas shrink the table size.
#define TPCC_SMALL					false
// Some of the transactions read the data but never use them. 
// If TPCC_ACCESS_ALL == fales, then these parts of the transactions
// are not modeled.
#define TPCC_ACCESS_ALL 			false 
#define WH_UPDATE					true
#define NUM_WH 						4
//
enum TPCCTxnType {TPCC_ALL, 
    TPCC_PAYMENT,
    TPCC_NEW_ORDER,
    TPCC_ORDER_STATUS,
    TPCC_DELIVERY,
    TPCC_STOCK_LEVEL};
extern TPCCTxnType 					g_tpcc_txn_type;

//#define TXN_TYPE					TPCC_ALL
#define PERC_PAYMENT 				0.5
#define FIRSTNAME_MINLEN 			8
#define FIRSTNAME_LEN 				16
#define LASTNAME_LEN 				16

#define DIST_PER_WARE				10

/***********************************************/
// TODO centralized CC management. 
/***********************************************/
#define MAX_LOCK_CNT				(20 * THREAD_CNT) 
#define TSTAB_SIZE                  50 * THREAD_CNT
#define TSTAB_FREE                  TSTAB_SIZE 
#define TSREQ_FREE                  4 * TSTAB_FREE
#define MVHIS_FREE                  4 * TSTAB_FREE
#define SPIN                        false

/***********************************************/
// Test cases
/***********************************************/
#define TEST_ALL					true
enum TestCases {
    READ_WRITE,
    CONFLICT
};
extern TestCases					g_test_case;
/***********************************************/
// DEBUG info
/***********************************************/
#define WL_VERB						true
#define IDX_VERB					false
#define VERB_ALLOC					true

#define DEBUG_LOCK					false
#define DEBUG_TIMESTAMP				false
#define DEBUG_SYNTH					false
#define DEBUG_ASSERT				false
#define DEBUG_CC					false //true

/***********************************************/
// Constant
/***********************************************/
// INDEX_STRUCT
#define IDX_HASH 					1
#define IDX_BTREE					2
// WORKLOAD
#define YCSB						1
#define TPCC						2
#define TEST						3
// Concurrency Control Algorithm
#define NO_WAIT						1
#define WAIT_DIE					2
#define DL_DETECT					3
#define TIMESTAMP					4
#define MVCC						5
#define HSTORE						6
#define OCC							7
#define TICTOC						8
#define SILO						9
#define VLL							10
#define HEKATON 					11
#define TWO_PL_SF                                       12
//Isolation Levels 
#define SERIALIZABLE				1
#define SNAPSHOT					2
#define REPEATABLE_READ				3
// TIMESTAMP allocation method.
#define TS_MUTEX					1
#define TS_CAS						2
#define TS_HW						3
#define TS_CLOCK					4

#endif
